terraform/eval_diff.go

```diff
@@ -256,12 +255,18 @@ func (n *EvalDiff) Eval(ctx EvalContext) (interface{}, error) {
 				))
 				continue
 			}
-			priorChangedVal, err := path.Apply(priorVal)
-			if err != nil {
-				// Should never happen since prior and changed should be of
-				// the same type, but we'll allow it for robustness.
-				reqRep.Add(path)
+
+			// make sure we have valid Values for both
+			switch {
+			case priorChangedVal == cty.NilVal && plannedChangedVal == cty.NilVal:
+				// this should never happen without ApplyPath errors above
+				panic("requires replace path returned 2 nil values")
+			case priorChangedVal == cty.NilVal:
+				priorChangedVal = cty.NullVal(plannedChangedVal.Type())
```

Comment:
In the case where the thing being changed contains `cty.DynamicPseudoType` attributes in its schema it is possible for the prior and planned to actually _not_ have the same type (contradicting the comment that used to be here, which was incorrect for a number of reasons :confounded: )

In think in this context it doesn't actually matter, since we're only using these values to run `Equals` below and that'll never show a null as equal to a non-null anyway, but maybe it's worth noting here that it's generally more correct to use the schema's implied type to create a null but we're intentionally not doing that here because we know it's safe, just because I worry that a future reader may read this and think it a good general usage example to copy elsewhere, in situations where the exact type of null may be important.

Reply:
Ah yes, that was intended here, since It was only for the comparison and I didn't want to traverse the schema to check for a dynamic type. I'll add the comment for our future selves so we don't rely on the type.